File: TypeScript/app/updateStrategies/conjuredStrategy.ts
import Item from "./../models/Item";
import IUpdateStrategy from "./IUpdateStrategy";

export default class ConjuredStrategy implements IUpdateStrategy {
    constructor(private qualityFactor : number) {}

    updateItem(item: Item) : Item {
        let quality = Math.max(0, item.quality - 2 * this.qualityFactor);
        let sellIn = item.sellIn - 1;
        return new Item(item.name, sellIn, quality);
    }
}

File: Java/src/main/java/com/gildedrose/ConjuredStrategy.java
package com.gildedrose;

public enum ConjuredStrategy implements QualityUpdateStrategy {
    INSTANCE;

    @Override
    public void applyTo(Item item) {
        if (item.sellIn > 0) {
            item.quality = Math.max(0, item.quality-2);
        } else {
            item.quality = Math.max(0, item.quality-4);
        }
        --item.sellIn;
    }
}
